C#/v1/NoarkWsClientSample/NoarkWsClientSample/DocumasterClients.cs

```diff
@@ -0,0 +1,95 @@
+using Documaster.WebApi.Client.Noark5.Client;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace NoarkWsClientSample
+{
+    public class DocumasterClients
+    {
+        protected NoarkClient noarkClient;
+        protected Oauth2HttpClient idpClient;
+
+        private string refreshToken;
+        private DateTime acccessTokenExpirationTime;
+        private Options opts;
+
+        public DocumasterClients(Options opts)
+        {
+            this.opts = opts;
+
+            /*
+             * Using the Noark 5 web services requires providing a valid access token.
+             * The way this token is obtained depends on the system implementing the services.
+             * This sample code obtains the token from the Documaster's identity provider service with the help of a designated Documaster IDP client.
+             * If the c# Noark client is used in the context of an application that has access to a web browser,
+             * we strongly recommend choosing the Oauth2 Authorization Code Grant Flow to avoid providing username and password.
+             * The c# Noark client provides both synchronous and 'async' methods.
+             * The 'async' methods should be called if the client is used in the context of a an application with an user interface.
+             * The synchronous methods can be used in console applications and libraries.
+             */
+
+            //Initialize an IDP client and request an authorization token
+            InitIdpClient(opts);
+
+            //Initialize a Noark client
+            InitNoarkClient(opts);
+        }
+
+        public NoarkClient getNoarkClient()
+        {
+            RefreshAccessToken();
+            return noarkClient;
+        }
+
+        private void RefreshAccessToken()
+        {
+            //access token expires in 60 minutes
+
+            if (refreshToken == null)
+            {
+                //PasswordGrantTypeParams needs the following information:
+                //ClientId: OAuth2 client_id
+                //ClientSecret: OAuth2 client_secret
+                //Username: The user's username
+                //Password: The user's password
+                //scope: "openid" in case of Documaster IDP
+                PasswordGrantTypeParams passwordGrantTypeParams = new PasswordGrantTypeParams(opts.ClientId,
+                    opts.ClientSecret, opts.Username, opts.Password, OpenIDConnectScope.OPENID);
+                AccessTokenResponse accessTokenResponse =
+                    idpClient.GetTokenWithPasswordGrantType(passwordGrantTypeParams);
+                acccessTokenExpirationTime = DateTime.Now.AddSeconds(accessTokenResponse.ExpiresInMs);
+                refreshToken = accessTokenResponse.RefreshToken;
+
+                noarkClient.AuthToken = accessTokenResponse.AccessToken;
+            }
+            else if (DateTime.Now > acccessTokenExpirationTime)
```

Comment:
Would it be wise to refresh the token a few seconds before it expires to account for clocks being a little bit out of sync and long network latency?

```suggestion
            else if (DateTime.Now > acccessTokenExpirationTime.AddSeconds(-20) )
```

Reply:
Thanks, fixed.